src/zenml/models/v2/base/base.py

```diff
@@ -0,0 +1,225 @@
+from pydantic.generics import GenericModel
+
+from zenml.exceptions import HydrationError
+from zenml.utils.pydantic_utils import YAMLSerializationMixin
+
+# -------------------- Base Model --------------------
+
+
+class BaseZenModel(YAMLSerializationMixin):
+    """Base model class for all ZenML models.
+
+    This class is used as a base class for all ZenML models. It provides
+    functionality for tracking analytics events and proper encoding of
+    SecretStr values.
+    """
+
+    ANALYTICS_FIELDS: ClassVar[List[str]] = []
+
+    def get_analytics_metadata(self) -> Dict[str, Any]:
+        """Get the analytics metadata for the model.
+
+        Returns:
+            Dict of analytics metadata.
+        """
+        metadata = {}
+        for field_name in self.ANALYTICS_FIELDS:
+            metadata[field_name] = getattr(self, field_name, None)
+        return metadata
+
+    class Config:
+        """Pydantic configuration class."""
+
+        # This is needed to allow the REST client and server to unpack SecretStr
+        # values correctly.
+        json_encoders = {
+            SecretStr: lambda v: v.get_secret_value()
+            if v is not None
+            else None
+        }
+
+        # Allow extras on all models to support forwards and backwards
+        # compatibility (e.g. new fields in newer versions of ZenML servers
+        # are allowed to be present in older versions of ZenML clients and
+        # vice versa).
+        extra = "allow"
+
+
+# -------------------- Request Model --------------------
+
+
+class BaseRequest(BaseZenModel):
+    """Base request model.
+
+    Used as a base class for all request models.
+    """
+
+
+# -------------------- Response Model --------------------
+
+
+class BaseResponseBody(BaseZenModel):
+    """Base body model.
+
+    Used as a base class for all body models associated with responses.
+    Features a creation and update timestamp.
+    """
+
+    created: Optional[datetime] = Field(
+        title="The timestamp when this resource was created.",
+        default=None,
+    )
+    updated: Optional[datetime] = Field(
+        title="The timestamp when this resource was last updated.",
+        default=None,
+    )
+
+
+class BaseResponseMetadata(BaseZenModel):
+    """Base metadata model.
+
+    Used as a base class for all metadata models associated with responses.
+    """
+
+
+AnyBody = TypeVar("AnyBody", bound=BaseResponseBody)
+AnyMetadata = TypeVar("AnyMetadata", bound=BaseResponseMetadata)
+
+
+class BaseResponse(GenericModel, Generic[AnyBody, AnyMetadata], BaseZenModel):
+    """Base domain model."""
+
+    id: UUID = Field(title="The unique resource id.")
+
+    # Body and metadata pair
+    body: "AnyBody" = Field(title="The body of the resource.")
+    metadata: Optional["AnyMetadata"] = Field(
+        title="The metadata related to this resource."
+    )
+
+    def get_hydrated_version(self) -> "BaseResponse[AnyBody, AnyMetadata]":
```

Comment:
I had a brief discussion with @schustmi about this. 

First of all, it needs to return a version of the same object with a populated metadata field instead of returning _just_ the metadata field. This is necessary down the line because we check the values within the body as well. 

As for the term `hydration`, I think the term is pretty well known within the team. If you would like to change this for external users, we need to rename many more methods and parameters not just in this PR but also in the RBAC PR. This is why I would discuss it in a separate thread and implement it if we decide to in a split fashion.